Add determinism developer documentation (#5718)
NVIDIA/Megatron-LM: docs/developer/determinism/op-catalog.md

---
orphan: true
---

> # Determinism op catalog

Reviewer: ```suggestion
# Determinism Operation Catalog
```

> > **Audience:** developers reviewing or extending deterministic-mode coverage.

Reviewer: ```suggestion
>This content is for developers reviewing or extending deterministic-mode coverage.
```

Reviewer: calling out things like Audience and reiterating where various items are in the content table of contents is not typical of professional technical content.

> > Terms are defined in the [glossary](./glossary.md). This page describes what

Reviewer: I am really not a fan of " This page describes what
> is merged on `main`; in-progress work is tracked in
> [issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785)."   

I strongly encourage you to delete it. 

If you really want to keep it, I would move it to the very end of this document.

> is merged on `main`; in-progress work is tracked in
> [issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785).

The catalog is organized into two buckets. The project goal is to keep shrinking

> the second bucket (operations deterministic mode cannot support yet) and to keep

Reviewer: ```suggestion
the second bucket (operations that deterministic mode cannot support yet) and to keep
```

making the first bucket (operations with a deterministic code path) faster.

> Most operations in a training step need no entry here at all: elementwise ops,

Reviewer: ```suggestion
Most operations in a training step need no entry here at all. Elementwise ops,
```

GEMMs under the pinned cuBLAS workspace, rank-indexed collectives (all-gather,
all-to-all, broadcast), stable sorts, and unique-index writes are deterministic
as-is. The tables list only the operations where a choice is made.

> ## 1. Operations with a deterministic code path

Reviewer: ```suggestion
## Deterministic Code Path Operations
```

Selected by `torch.are_deterministic_algorithms_enabled()` or

> `config.deterministic_mode`; the default-mode path stays in the other branch.

Reviewer: ```suggestion
`config.deterministic_mode`. The default-mode path stays in the other branch.
```

> | Operation | Where | Deterministic path | Default path |

Reviewer: ```suggestion
| Operation | Where | Deterministic Path | Default Path |
```

| --- | --- | --- | --- |
| MoE token unpermute (combine) | `megatron/core/transformer/moe/moe_utils.py:517` | `index_add_` — deterministic under torch deterministic algorithms and CUDA-graph safe | `scatter_add_` (atomic accumulation) |

> | MoE routing map / probabilities | `megatron/core/transformer/moe/moe_utils.py:823` | `index_put_(accumulate=False)` row-wise writes | out-of-place `scatter` |

Reviewer: ```suggestion
| MoE routing map and probabilities | `megatron/core/transformer/moe/moe_utils.py:823` | `index_put_(accumulate=False)` row-wise writes | out-of-place `scatter` |
```

| Vocab-parallel embedding | `megatron/core/tensor_parallel/layers.py:299` | direct indexing `weight[idx]` (deterministic backward) | `F.embedding` (non-deterministic atomic backward) |
| Gated-delta-net kernel | `megatron/core/ssm/gated_delta_net.py:212` | torch `chunk_gated_delta_rule` | FLA fused kernel |
| Gated-delta-net causal conv1d | `megatron/core/ssm/gated_delta_net.py:429` | `F.conv1d` (plus transposes) | FLA `causal_conv1d` |
| Mamba/SSM Triton ops | `megatron/core/ssm/ops/determinism.py` | one fixed autotune config plus a zero-initialized tiled workspace reduced with an ordered `sum` | timing-based autotune, uninitialized workspace |
| Transformer Engine attention | `megatron/core/extensions/transformer_engine.py:1698` | requires `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`, under which TE selects only backends that support deterministic execution (including deterministic FlashAttention backward) | TE picks freely, including atomic-accumulation attention backward |

> | Inference DP scheduling / RL rollout order | `megatron/core/inference/engines/dynamic_engine.py`, `megatron/rl/rl_utils.py` | sort by stable key | completion order |

Reviewer: ```suggestion
| Inference DP scheduling and RL rollout order | `megatron/core/inference/engines/dynamic_engine.py`, `megatron/rl/rl_utils.py` | sort by stable key | completion order |
```

> Environment controls that make the rest of the step deterministic (validated

Reviewer: ```suggestion
The following environment controls make the rest of the step deterministic.
The flag `--deterministic-mode` validates and defaults these settings. For
details, refer to `megatron/training/determinism.py`.

- `NCCL_ALGO=Ring`. Tree is rejected because its reduction order is not
  user-controllable.
- `CUBLAS_WORKSPACE_CONFIG=:4096:8` (or `:16:8`).
- `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`.
- `MAMBA_DETERMINISTIC` must not be disabled.
```

Reviewer: then delete lines 38-42

and defaulted by `--deterministic-mode`; see
`megatron/training/determinism.py`): `NCCL_ALGO=Ring` (Tree is rejected — its
reduction order is not user-controllable), `CUBLAS_WORKSPACE_CONFIG=:4096:8`
(or `:16:8`), `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`, and `MAMBA_DETERMINISTIC`
must not be disabled.

> ## 2. Operations without determinism support

Reviewer: ```suggestion
## Non-Deterministic Code Path Operations
```

Deterministic mode either rejects these at validation (fails closed) or they
are known open gaps.

> | Operation / feature | Where enforced or observed | Status |

Reviewer: ```suggestion
| Operation or Feature | Where Enforced or Observed | Status |
```

| --- | --- | --- |

> | Fused cross-entropy loss (`--cross-entropy-loss-fusion`) | rejected by `--deterministic-mode` (`megatron/training/determinism.py`) | The fused kernel is non-deterministic. Open question whether a deterministic variant is feasible; until then the native vocab-parallel path is used. |

Reviewer: ```suggestion
| Fused cross-entropy loss (`--cross-entropy-loss-fusion`) | rejected by `--deterministic-mode` (`megatron/training/determinism.py`) | The fused kernel is non-deterministic. Whether a deterministic variant is feasible remains an open question. Until then, the framework uses the
native vocab-parallel path. |
```

| TP communication overlap (`--tp-comm-overlap`) | rejected by `--deterministic-mode` | Overlapped collective ordering is not reproducible. |
| Packed sequence (`thd`) in gated-delta-net | `megatron/core/ssm/gated_delta_net.py:313` assert | No deterministic packed-sequence SSM path exists yet. |

> | Cross-allocation floating-point collectives (TP all-reduce, DP grad reduce-scatter) | open gap | `NCCL_ALGO=Ring` pins the algorithm but not the physical ring an allocation gets, so bit-exactness across *different* allocations is not guaranteed by the env pin alone for these reductions. Runs repeated within one allocation, or on allocations with identical topology, are covered. |

Reviewer: ```suggestion
| Cross-allocation floating-point collectives (TP all-reduce, DP grad reduce-scatter) | open gap | `NCCL_ALGO=Ring` pins the algorithm but not the physical ring an
allocation receives. The environment variable alone does not guarantee
bit-exactness across *different* allocations for these reductions. Runs
repeated within one allocation, or on allocations with identical topology,
remain bit-exact. |
```

> ## 3. Performance notes

Reviewer: ```suggestion
## Performance Notes
```

> The deterministic paths above cost roughly 15% step time versus default mode,

Reviewer: ```suggestion
The deterministic paths above cost roughly 15% of step time compared to
default mode, which varies by model. MoE-heavy models pay more. Measured
examples range from approximately 4% on a large dense model to
approximately 17% on a hybrid MoE model. The measured hotspots are the
deterministic MoE scatter and unpermute path, the sorted router top-k,
attention backward, and grouped-GEMM wgrad.

Reducing this cost is a tracked workstream in
`issue #5785 <https://github.com/NVIDIA/Megatron-LM/issues/5785>`_. A
change to any row above needs a bit-exact test and a deterministic compared
to default performance comparison test
(`tests/performance_tests/shell_test_utils/determinism/`).
```

Reviewer: then delete 59-65

---
orphan: true
---

# Determinism Operation Catalog

> This content is for developers reviewing or extending deterministic-mode
> coverage. Terms are defined in the [glossary](./glossary.md).

> The catalog is organized into two buckets. The project goal is to keep shrinking

Reviewer: ```suggestion
The catalog has two buckets:

- Operations with a deterministic code path
- Operations that deterministic mode cannot support yet

The project goal is to shrink the second bucket and make the first bucket
faster.

```

Author: Applied in 257e0ebf2.

> the second bucket (operations that deterministic mode cannot support yet) and to keep

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> making the first bucket (operations with a deterministic code path) faster.

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> Most operations in a training step need no entry here at all. Elementwise ops,

Reviewer: ```suggestion
Most operations in a training step need no entry here. The following are
deterministic as-is:

- Elementwise ops
- GEMMs under the pinned cuBLAS workspace
- Rank-indexed collectives (all-gather, all-to-all, broadcast)
- Stable sorts
- Unique-index writes

The tables list only the operations where a choice is made.
```

Author: Applied in 257e0ebf2.

> GEMMs under the pinned cuBLAS workspace, rank-indexed collectives (all-gather,

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> all-to-all, broadcast), stable sorts, and unique-index writes are deterministic

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> as-is. The tables list only the operations where a choice is made.

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

| MoE routing map and probabilities | `megatron/core/transformer/moe/moe_utils.py:823` | `index_put_(accumulate=False)` row-wise writes | out-of-place `scatter` |
| Vocab-parallel embedding | `megatron/core/tensor_parallel/layers.py:299` | direct indexing `weight[idx]` (deterministic backward) | `F.embedding` (non-deterministic atomic backward) |
| Gated-delta-net kernel | `megatron/core/ssm/gated_delta_net.py:212` | torch `chunk_gated_delta_rule` | FLA fused kernel |
| Gated-delta-net causal conv1d | `megatron/core/ssm/gated_delta_net.py:429` | `F.conv1d` (plus transposes) | FLA `causal_conv1d` |
| Mamba/SSM Triton ops | `megatron/core/ssm/ops/determinism.py` | one fixed autotune config plus a zero-initialized tiled workspace reduced with an ordered `sum` | timing-based autotune, uninitialized workspace |
| Transformer Engine attention | `megatron/core/extensions/transformer_engine.py:1698` | requires `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`, under which TE selects only backends that support deterministic execution (including deterministic FlashAttention backward) | TE picks freely, including atomic-accumulation attention backward |

> | Inference DP scheduling and RL rollout order | `megatron/core/inference/engines/dynamic_engine.py`, `megatron/rl/rl_utils.py` | sort by stable key | completion order |

Reviewer 2: Is there a better alternative to referencing the line number here, since it may change over time? We could use a permanent link to the specific revision or simply omit the line number.

The following environment controls make the rest of the step deterministic.
The flag `--deterministic-mode` validates and defaults these settings. For
details, refer to `megatron/training/determinism.py`.

> - `NCCL_ALGO=Ring`. Tree is rejected because its reduction order is not

Reviewer: ```suggestion
- `NCCL_ALGO=Ring`. The tree is rejected because its reduction order is not
```

Author: Applied in 257e0ebf2.

| Operation or Feature | Where Enforced or Observed | Status |
| --- | --- | --- |
| Fused cross-entropy loss (`--cross-entropy-loss-fusion`) | rejected by `--deterministic-mode` (`megatron/training/determinism.py`) | The fused kernel is non-deterministic. Whether a deterministic variant is feasible remains an open question. Until then, the framework uses the native vocab-parallel path. |
| TP communication overlap (`--tp-comm-overlap`) | rejected by `--deterministic-mode` | Overlapped collective ordering is not reproducible. |
| Packed sequence (`thd`) in gated-delta-net | assertion in `megatron/core/ssm/gated_delta_net.py` | No deterministic packed-sequence SSM path exists yet. |
| Cross-allocation floating-point collectives (TP all-reduce, DP grad reduce-scatter) | open gap | `NCCL_ALGO=Ring` pins the algorithm but not the physical ring an allocation receives. The environment variable alone does not guarantee bit-exactness across *different* allocations for these reductions. Runs repeated within one allocation, or on allocations with identical topology, remain bit-exact. |

## Performance Notes

The deterministic paths above cost roughly 15% of step time compared to

> default mode, which varies by model. MoE-heavy models pay more. Measured

Reviewer: ```suggestion
default mode, which varies by model. Models that rely heavily on Mixture of
Experts (MoE) pay more. Measured examples range from approximately four percent on a large dense
model to approximately 17% on a hybrid MoE model. The measured hotspots are:

- The deterministic MoE scatter and unpermute path
- The sorted router top-k
- Attention backward
- Grouped-GEMM weight gradient (wgrad)
```

Author: Applied in 257e0ebf2.

> examples range from approximately 4% on a large dense model to approximately

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> 17% on a hybrid MoE model. The measured hotspots are the deterministic MoE

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> scatter and unpermute path, the sorted router top-k, attention backward, and

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

> grouped-GEMM wgrad.

Reviewer: ```suggestion
```

Author: Applied in 257e0ebf2.

Reducing this cost is a tracked workstream in
[issue #5785](https://github.com/NVIDIA/Megatron-LM/issues/5785). A change to

> any row above needs a bit-exact test and a deterministic-vs-default performance

Reviewer: ```suggestion
any row above needs a bit-exact test and a comparison of deterministic and
```

Author: Applied in 257e0ebf2.

> comparison (`tests/performance_tests/shell_test_utils/determinism/`).

Reviewer: ```suggestion
default performance (`tests/performance_tests/shell_test_utils/determinism/`).
```

Author: Applied in 257e0ebf2.
